File: 백준/Gold/2623. 음악프로그램/음악프로그램.java
import java.util.*;
import java.io.*;

class Main {
    static int n, m, num;
    static ArrayList<Integer>[] child;
    static int[] degree, order;
    static Queue<Integer> q;
    static ArrayList<Integer> answer;
    
    public static void main(String[] args) throws IOException  {
        BufferedReader bf = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(bf.readLine());

        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());

        child = new ArrayList[n+1];
        q = new ArrayDeque<>();
        answer = new ArrayList<>();
        degree = new int[n+1];

        for (int i = 1; i < n+1; i++){
            child[i] = new ArrayList<>();
        }

        for (int i = 0; i < m; i++){
            st = new StringTokenizer(bf.readLine());
            num = Integer.parseInt(st.nextToken());
            order = new int[num];

            for (int j = 0; j < num; j++){
                order[j] = Integer.parseInt(st.nextToken());
            }

            for (int j = 0; j < num-1; j++){
                child[order[j]].add(order[j+1]);
                degree[order[j+1]]++;
            }
        }

        for (int i = 1; i <= n; i++) {
            if (degree[i] == 0) {
                q.add(i);
            }
        }

        while (!q.isEmpty()) {
            int x = q.poll();
            answer.add(x);

            for (int num : child[x]) {
                degree[num]--;
                if (degree[num] == 0) {
                    q.add(num);
                }
            }
        }

        if (answer.size() < n) {
            System.out.println(0);
        } else {
            for (int a : answer) {
                System.out.println(a);
            }
        }
    }
}

File: 백준/Gold/2623. 음악프로그램/음악프로그램.py
from collections import deque

n, m = map(int, input().split())
child = [[] for _ in range(n + 1)]
degree = [0] * (n + 1)
q = deque()
answer = []
for _ in range(m):
    num, *order = list(map(int, input().split()))
    for i in range(len(order) - 1):
        child[order[i]].append(order[i + 1])
        degree[order[i + 1]] += 1

for i in range(1, n + 1):
    if degree[i] == 0:
        q.append(i)

while q:
    x = q.popleft()
    answer.append(x)

    for num in child[x]:
        degree[num] -= 1
        if degree[num] == 0:
            q.append(num)

if len(answer) < n:
    print(0)
else:
    for a in answer:
        print(a)
